Update design docs with syntax changes from #2760 (#2866)
carbon-language/carbon-lang: docs/design/generics/details.md

class Point {
  var x: f64;
  var y: f64;
  extend impl as Vector {
    fn Add[self: Self](b: Self) -> Self { ... }
    fn Scale[self: Self](v: f64) -> Self { ... }
  }
  extend impl as Drawable {
    fn Draw[self: Self]() { ... }
  }
}
```

In this case, all the functions `Add`, `Scale`, and `Draw` end up a part of the
API for `Point`. This means you can't implement two interfaces that have a name
in common (unless you use an `impl` without `extend` for one or both, as

> described below).

Reviewer: ```suggestion
in common (unless you use an `impl` without `extend` for one or both, for
an [external impl](#external-impl)).
```

A quirk of this change is that it's hard to search on now, so given it's not the immediately following text, maybe a link to be explicit?

Given a value of type `Point3` and an interface `Vector` implemented for that
type, you can access the methods from that interface using a
[qualified member access expression](terminology.md#qualified-member-access-expression)
whether or not the implementation is done externally with an `impl` declaration

> without `extend`. The qualified member access expression writes the member's

Reviewer: ```suggestion
whether or not the implementation is done with an [external `impl` declaration](#external-impl). The qualified member access expression writes the member's
```

Just a suggestion -- the "with ... without" could be difficult to parse, and linking provides context.

}
```

> This is equivalent to naming the type between `impl` and `as`, though this form

Reviewer: This is referred to as the "existing type", line 386. "type to the left of `as`" on 3788.

I think it'd be good to refer to it using a consistent phrase, and maybe there should be a name for it, such as "`impl` receiver"? (e.g., "This is equivalent to specifying the `impl` receiver, though this form ...")

Author: According to https://doc.rust-lang.org/reference/items/implementations.html , this is called the "implementing type" in Rust.

Reviewer: I don't really have a strong opinion here if others find that logical, although I think "implementing type" is confusing for external impls, where I wouldn't think of the type that's targeted by the external impl as the type providing implementation.

This is pre-existing though, so I want to be clear I don't think it should block merging.

    if the element type is comparable.
-   A container is copyable if its elements are.

> This may be done with an [external `impl`](#external-impl) by specifying a

Reviewer: In theory, an external `impl` could be inline, and that's strongly implied by saying "more specific `Self`" (which is invalid before `as` when out-of-line). However, the example is out-of-line. If inline external `impl` is supported for this, perhaps the example should be changed? Otherwise, I think the prior text for "out-of-line" may have been more precise.

(an inline definition might not work in the `forall` because the `Vector(T)` would be incomplete)

Reviewer: As I'm glancing at this again -- a later example clarifies it's intended to work, so maybe just adjust the example to show inline?

Author: - I made the change to make it more accurate, inline external `impl` is supported for this.
- I will delete the word `Self` since it seems to be causing some confusion. It is true that you can't put `Self` before `as`, but I was referring to it as the `Self` type since `Self` inside the `impl` definition body is set to that type.
- External impls can be declared inline, but they will most commonly be out-of-line, and the example reflected that. There is an example further down to show it can also be inline.

Reviewer: Thanks! I agree removing `Self` clarifies.

> more-specific `Self` type to the left of the `as` in the declaration:

Reviewer: ```suggestion
more specific `Self` type to the left of the `as` in the declaration:
```

I don't think this should be hyphenated.

}
```

Note that no `forall` clause or type may be specified when declaring an `impl`
with the `extend` keyword:

```
class Array(T:! type, template N:! i64) {
  // ❌ Illegal: nothing allowed before `as` after `extend impl`:
  extend impl forall [P:! Printable] Array(P, N) as Printable { ... }
}
```

> **Note:** This was changed in
> [proposal #2760](https://github.com/carbon-language/carbon-lang/pull/2760).
> 

Reviewer: ```suggestion
```

I assume this was temporary while #2760 was in flight, since references are generally at the bottom.

Author: I think this comment is useful here in context, rather than somewhere else.

Instead, the class can declare aliases to members of the interface. Those
aliases will only be usable when the type implements the interface.

```
class Array(T:! type, template N:! i64) {
  alias Print = Printable.Print;
}
impl forall [P:! Printable] Array(P, N) as Printable { ... }

impl String as Printable { ... }
var can_print: Array(String, 2) = ("Hello ", "world");
// ✅ Allowed: `can_print.Print` resolves to
// `can_print.(Printable.Print)`, which exists as long as
// `Array(String, 2) impls Printable`, which exists since
// `String impls Printable`.
can_print.Print();

var no_print: Array(Unprintable, 2) = ...;
// ❌ Illegal: `no_print.Print` resolves to
// `no_print.(Printable.Print)`, but there is no
// implementation of `Printable` for `Array(Unprintable, 2)`
// as long as `Unprintable` doesn't implement `Printable`.
no_print.Print();
```

> It is still legal to declare or define an external impl lexically with the class

Reviewer: ```suggestion
It is still legal to declare or define an external impl lexically inside the class
```

scope, as in:

```
class Array(T:! type, template N:! i64) {
  // ✅ Allowed: external impl defined in class scope may use `forall`

>   // and/or specify a type.

Reviewer: ```suggestion
  // or specify a type.
```

Style encourages rewriting to avoid `and/or`. In this case, `impl X forall Y as Z` is invalid, so maybe we can drop `and`? ("may specify a type or use `forall`" might read more easily, too)
https://developers.google.com/style/word-list#and-or

Author: `impl X forall Y as Z` is never legal, but that is because the `forall Y` comes first (and should include `[`...`]`). `impl forall [Y] X as Z` is definitely a thing, as show by this specific example. I've updated the text to replace "and/or."

```
interface Foo(T:! type) { ... }
class Pair(T:! type, U:! type) { ... }
impl forall [T:! type] Pair(T, T) as Foo(T) { ... }
```

You may also define the `impl` inline, but it may not be combined with the

> `extend` keyword:

Reviewer: Could 3847-3856 and 3875-3882 be unified, since they are both explaining that an `impl` can be inline without `extend`?

Author: I've added the words "as before" to make it clear this is just another example of the same thing.

Reviewer: Just to note, it still feels redundant to me as both examples seem like they're demonstrating an absence of `extend`, but maybe others will find the repetition useful.
